### ISSUE TITLE
Missing rd_atomic??_init()s?

### ISSUE THREAD

**Issue by alangcarter** [edenhill/librdkafka#1813]
Description
===========
I am upgrading from C API 0.9.5 to 0.11.4, building librdkafka on Windows 7 using VS2017.

At 0.9.5 there were missing rd_atomic64_init() calls. Without them, the first call to e.g. 

rd_atomic64_add(&rktrans->rktrans_rkb->rkb_c.wakeups, 1);

in rd_kafka_transport_poll() would throw because of an uninitialized mtx_t.

Now at 0.11.4 it looks like the initializations are still missing, e.g.

$ grep wakeups *.[ch]
rdkafka.c:                           "\"wakeups\":%"PRIu64", "
rdkafka.c:                           rd_atomic64_get(&rkb->rkb_c.wakeups),
rdkafka_broker.c:                 * rkb_ops queue wakeups are always enabled though,
rdkafka_broker.h:                rd_atomic64_t wakeups;       /* Poll wakeups */
rdkafka_transport.c:        rd_atomic64_add(&rktrans->rktrans_rkb->rkb_c.wakeups, 1);

That's a dump, a get(), a comment, a declaration and an add(). No rd_atomic64_init() to zero the counter or initialize the mtx_t.

When I was stuck at 0.9.5 (messy real world historical issues) the problem was so historical it wasn't worth mentioning. With the rd_atomic64_init()s still missing at 0.11.4 I'm wondering if I am misunderstanding something, or if I should apply the init()s to 0.11.4 and submit a pull request.

Checklist
=========

librdkafka v0.11.4

Apache Kafka v0.10.1 (connection not necessary)

Producer: Windows 7, VS2017 Brokers: Centos 7, gcc 4.8.5 (brokers not involved)

Global configuration: api.version.fallback.ms=1200000
Global configuration: api.version.request=true
Global configuration: auto.commit.interval.ms=5000
Global configuration: batch.num.messages=2000
Global configuration: broker.address.family=any
Global configuration: broker.address.ttl=1000
Global configuration: broker.version.fallback=0.9.0
Global configuration: builtin.features=snappy,ssl,sasl,regex,lz4
Global configuration: check.crcs=false
Global configuration: client.id=rdkafka
Global configuration: compression.codec=none
Global configuration: coordinator.query.interval.ms=600000
Global configuration: delivery.report.only.error=false
Global configuration: dr_msg_cb=0136F510
Global configuration: enable.auto.commit=true
Global configuration: enable.auto.offset.store=true
Global configuration: enable.partition.eof=true
Global configuration: enabled_events=0
Global configuration: error_cb=0136F5D0
Global configuration: fetch.error.backoff.ms=500
Global configuration: fetch.message.max.bytes=1048576
Global configuration: fetch.min.bytes=1
Global configuration: fetch.wait.max.ms=100
Global configuration: group.protocol.type=consumer
Global configuration: heartbeat.interval.ms=1000
Global configuration: internal.termination.signal=0
Global configuration: log.connection.close=true
Global configuration: log.queue=false
Global configuration: log.thread.name=true
Global configuration: log_cb=0136F620
Global configuration: log_level=6
Global configuration: max.in.flight.requests.per.connection=1000000
Global configuration: message.copy.max.bytes=65535
Global configuration: message.max.bytes=10000000
Global configuration: message.send.max.retries=0
Global configuration: metadata.broker.list=10.32.0.104:6667,10.32.0.104:6668,10.32.0.104:6669
Global configuration: metadata.max.age.ms=-1
Global configuration: metadata.request.timeout.ms=60000
Global configuration: offset.store.method=broker
Global configuration: opaque=006BDC88
Global configuration: open_cb=0132EE20
Global configuration: partition.assignment.strategy=range,roundrobin
Global configuration: queue.buffering.max.kbytes=4000000
Global configuration: queue.buffering.max.messages=50000
Global configuration: queue.buffering.max.ms=200
Global configuration: queued.max.messages.kbytes=1000000
Global configuration: queued.min.messages=100000
Global configuration: receive.message.max.bytes=100000000
Global configuration: reconnect.backoff.jitter.ms=500
Global configuration: retry.backoff.ms=100
Global configuration: sasl.kerberos.principal=kafkaclient
Global configuration: sasl.kerberos.service.name=kafka
Global configuration: sasl.mechanisms=GSSAPI
Global configuration: security.protocol=plaintext
Global configuration: session.timeout.ms=30000
Global configuration: socket.blocking.max.ms=1000
Global configuration: socket.keepalive.enable=false
Global configuration: socket.max.fails=3
Global configuration: socket.nagle.disable=false
Global configuration: socket.receive.buffer.bytes=0
Global configuration: socket.send.buffer.bytes=0
Global configuration: socket.timeout.ms=60000
Global configuration: socket_cb=01308510
Global configuration: statistics.interval.ms=0
Global configuration: stats_cb=0136F6B0
Global configuration: topic.metadata.refresh.fast.cnt=10
Global configuration: topic.metadata.refresh.fast.interval.ms=250
Global configuration: topic.metadata.refresh.interval.ms=300000
Global configuration: topic.metadata.refresh.sparse=true

Producer topic configuration: auto.commit.enable=true
Producer topic configuration: auto.commit.interval.ms=60000
Producer topic configuration: auto.offset.reset=largest
Producer topic configuration: compression.codec=inherit
Producer topic configuration: consume.callback.max.messages=0
Producer topic configuration: message.timeout.ms=60000
Producer topic configuration: offset.store.method=broker
Producer topic configuration: offset.store.path=.
Producer topic configuration: offset.store.sync.interval.ms=-1
Producer topic configuration: partitioner_cb=0136F670
Producer topic configuration: produce.offset.report=false
Producer topic configuration: request.required.acks=-1
Producer topic configuration: request.timeout.ms=60000

**Comment by edenhill** [comment]
There are a couple of places calling rd_atomic.._init(), but far from all that use atomics.

However, this isn't required on windows since the rd_atomic*s are backed by InterlockedAdd/Exchange() which to my knowledge does not require initialization, there are no mutexes being used.

**Comment by alangcarter** [comment]
Now I see. I had to backport 0.9.5 to VS2012 (large build system, slow to upgrade), where InterlockedAdd64() was not available. So I had cut the _MSC_VER option out and was actually exercising the mtx_lock() code. That was how I discovered the missing init()s.

It's only the Windows code that needs the underlying CriticalSection in mtx_lock() to be initialized, so in practice the problem cannot happen with the up to date compiler. So there's no need to do the init()s, although they have their own #defines so it doesn't hurt.

Thanks for the prompt response - this mystery is now solved!
